MDEV-28153: Debian autobake- use absolute dependencies

grooverdan opened the pull request:
<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling this template <3

If you have any questions related to MariaDB or you just want to
hang out and meet other community members, please join us on
https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-28153*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed, what was it looking like before
   the change and how it's looking with this patch applied
3. Do you think this patch might introduce side-effects in
   other parts of the server?
-->
## Description

With autobake-deb.sh no doing explicit dependencies, we need to run this
explicitly as a prep step in salsa-ci so the dependencies can be pulled
before the real compilation begins.

Resolves 10.5 part of salsa in #2039.

## How can this PR be tested?

Currently running on https://salsa.debian.org/grooverdan/mariadb-server/-/pipelines/361865)

(10.8 test timed out)

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [X] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

<!--
You might consider answering some questions like:
1. Does this affect the on-disk format used by MariaDB?
2. Does this change any behavior experienced by a user
   who upgrades from a version prior to this patch?
3. Would a user be able to start MariaDB on a datadir
   created prior to your fix?
-->
## Backward compatibility
TODO: fill details here, if applicable, or remove the section

ottok reviewed (changes requested):
I don't understand the commit message / documentation. Why are these changes needed? What is broken - what do these changes fix?

What is _early bit of autobake-deb.sh_ that needs lsb-release? What is this SALSA_PREP thing? Why is columnstore stanza refactored?

ottok commented:
Changed
```
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ X ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*
```
to
```
## Basing the PR against the correct MariaDB version
- [X] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*
```

grooverdan commented:
> I don't understand the commit message / documentation. Why are these changes needed?

Since MDEV-28153, debian dependencies are now a property of the script. Without adjusting control before `apt-get build-dep`, an incorrect, possibly impossible, set of dependencies are generated.
 
> What is broken - what do these changes fix?

By running the prepare only of autobake-deb.sh, the control file has its build dependencies updated.  Running `apt-get build-dep` on any system will now pull the correct acheivable dependencies.

> What is _early bit of autobake-deb.sh_ that needs lsb-release?

`CODENAME="$(lsb_release -sc)"`

> What is this SALSA_PREP thing?

A flag to run the prepare stage only, so that `control`  (and `rules`) files only are updated.

> Why is columnstore stanza refactored?

It need to be id-potent so that a second run to trigger the build doesn't get two sets of included columnstore files, especially in `debian/control`

ottok commented:
> > I don't understand the commit message / documentation. Why are these changes needed?
> 
> Since MDEV-28153, debian dependencies are now a property of the script. Without adjusting control before `apt-get build-dep`, an incorrect, possibly impossible, set of dependencies are generated.
> 
> > What is broken - what do these changes fix?
> 
> By running the prepare only of autobake-deb.sh, the control file has its build dependencies updated. Running `apt-get build-dep` on any system will now pull the correct acheivable dependencies.

There is no need to run *prepare only*, either you run full autobake-deb.sh or you do a native build. A change based on the above assumption is mislead.

> > What is _early bit of autobake-deb.sh_ that needs lsb-release?
> 
> `CODENAME="$(lsb_release -sc)"`

This command works all the time and is part of standard Debian base. Why add it now?

This PR would be a lot more convincing if it referred a failing CI run on mainline and then a successful one from this PR. That is not the case now.
 
> > What is this SALSA_PREP thing?
> 
> A flag to run the prepare stage only, so that `control` (and `rules`) files only are updated.

Does not make sense as explained above.

> > Why is columnstore stanza refactored?
> 
> It need to be id-potent so that a second run to trigger the build doesn't get two sets of included columnstore files, especially in `debian/control`

This whole section is obsolete now when you changed/broke the previous columnstore packaging structure via ec62f46a612b52aa75d61b971d0a7eb781cc19c1. Either it is part of the native packaging (withfiles in debian/* and stanza in debian/control), or it is not. Any attempt for middle ground just causes a mess and is brings no benefit as far as I can see.

grooverdan commented:
> > > I don't understand the commit message / documentation. Why are these changes needed?
> > 
> > Since MDEV-28153, debian dependencies are now a property of the script. Without adjusting control before `apt-get build-dep`, an incorrect, possibly impossible, set of dependencies are generated.
> > > What is broken - what do these changes fix?
> > 
> > 
> > By running the prepare only of autobake-deb.sh, the control file has its build dependencies updated. Running `apt-get build-dep` on any system will now pull the correct acheivable dependencies.
> 
> There is no need to run _prepare only_, either you run full autobake-deb.sh or you do a native build. A change based on the above assumption is mislead.
> 
> > > What is _early bit of autobake-deb.sh_ that needs lsb-release?
> > 
> > 
> > `CODENAME="$(lsb_release -sc)"`
> 
> This command works all the time and is part of standard Debian base. Why add it now?
> 
> This PR would be a lot more convincing if it referred a failing CI run on mainline and then a successful one from this PR. That is not the case now.

https://salsa.debian.org/grooverdan/mariadb-server/-/jobs/2603984

> This whole section is obsolete now when you changed/broke the previous columnstore packaging structure via [ec62f46](https://github.com/MariaDB/server/commit/ec62f46a612b52aa75d61b971d0a7eb781cc19c1).

I noticed and corrected this 2885fb0ee496285b49efea7d84c6cd0a8e819319

I'm sorry we have differing opinions on the MDEV-28153 changes.

ottok commented:
Thanks for https://github.com/MariaDB/server/commit/2885fb0ee496285b49efea7d84c6cd0a8e819319. Please in the future ask for review for all commits before you push them. The PR where you pushed that commit you added those commits after review and proceeded to merge without waiting for second review of the updated version.

I see in https://salsa.debian.org/grooverdan/mariadb-server/-/jobs/2603984  `debian/autobake-deb.sh: line 79: lsb_release: command not found` but this commit is on this PR and I am against this PR, so if this is not merged, the failing build is not introduced in the first place.

To iterate, Salsa-CI builds on branches 10.5/6/7/8/9 are not broken and do not need this fix.

> I'm sorry we have differing opinions on the MDEV-28153 changes.

I still think that the claim about "apt cache being broken" is based on misunderstanding or flawed assumptions. The apt cache is never broken if the builder is healthy and capable of installing new packages. However, since you now refactored that, you should take ownership of the autobake-deb.sh contents. You can do that without having to pollute the salsa-ci.yml file with (at least so far seemingly) unnecessary `SALSA_PREP=1 debian/autobake-deb.sh` lines.

illuusio commented on debian/autobake-deb.sh:
```diff
@@ -91,6 +94,11 @@ case "${CODENAME}" in
     exit 1
 esac
 
+if [ -n "${SALSA_PREP:-}" ]
```
There could be some comment why this is here as I don't understood reasoning? If we are not in Salsa exit?

grooverdan replied:
`:-` to give it a default value in case its not set. Will prevent failures in bash strict `+x` mode and gets recognised by shellcheck.

ottok replied:
Please add inline comment on what this section is, what uses it and why. Anybody can look up the syntax on a man page but it is hard to guess why you have this seemingly unnecessary logic here.

ottok replied:
Ok, so there should be an inline comment saying that the section is tailored for use with https://github.com/MariaDB/mariadb.org-tools/blob/master/buildbot.mariadb.org/ci_build_images/debian.Dockerfile to make Docker images with preinstalled build-dependencies for each Debian/Ubuntu release, and thus you want to generate the `debian/control` files but not run autobake.

grooverdan commented:
Removed salsa line, Added the one dependency line only.

https://salsa.debian.org/grooverdan/mariadb-server/-/commit/35cb5b446cbbffa6729e5d55d81c224ed56d06d9/pipelines?ref=bb-10.5-danielblack-MDEV-28153-deb-explict-deps-salsa-postfix is the build in progress.

Matched up with https://github.com/MariaDB/mariadb.org-tools/pull/130 build.

ottok commented:
> https://salsa.debian.org/grooverdan/mariadb-server/-/commit/35cb5b446cbbffa6729e5d55d81c224ed56d06d9/pipelines?ref=bb-10.5-danielblack-MDEV-28153-deb-explict-deps-salsa-postfix is the build in progress

Yes, but I am asking where is the build that failed? The addition you did to salsa-ci.yml is unnecessary, it does not do anything. Referencing a build that passes does not prove anything about that that addition fixed (and it didn't fix anything, salsa-ci on 10.5 is already passing).

illuusio commented:
> The addition you did to salsa-ci.yml is unnecessary, it does not do anything.

Why does it not do anything? Is lsb already Salsa-CI dependency? I think point is that it's not breaking anything it's not that it should make it pass?
